---
title: "SpliceFlow"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SpliceFlow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
require(dplyr)
```

```{r setup}
library(SpliceFlow)
library(GenomicAlignments)
library(Rsamtools)
library(SummarizedExperiment)
```

## Build intron database
The function `prepareFlowRegions()` will create a an object containing all introns from the genome and the intron-exon pairs that will quantify splicing efficiency mentioned above. In order to construct an accurate intron database a junction file is required to pre-filter not used intron-exon that can imapct the analyasis

```{r}
ref <- system.file("data", "protein_coding.genes.gtf.gz", package = "SpliceFlow")
data("neuronal.junctions", package = "SpliceFlow")
regionsFlow <- prepareFlowRegions(ref, neuronal_junctions )
show(regionsFlow)
```

# Quantify 

```{r}
pathBams <- list.files(system.file("data", package = "SpliceFlow"), 
                       pattern = "\\.bam$", full.names = TRUE)
bamFiles = Rsamtools::BamFileList(pathBams)
names(bamFiles) <- c("nascent_R1", "nascent_R2", "polyA_R1", "polyA_R2")
# regions to quantify
# counts 
se = GenomicAlignments::summarizeOverlaps(regionsFlow@quantFlow,
                       bamFiles,
                       singleEnd = TRUE,
                       ignore.strand = TRUE)
rowData(se)$tx_name <- paste0( rowData(se)$gene_id, ":E", rowData(se)$exon_id) 
rowData(se)$transcript_id <- rowData(se)$exon_intron_region
colData(se)$condition = factor(c("nascent", "nascent", "mrna", "mrna"))
result <- SpliceFlow(se, regionsFlow)
# results stored in rowData
rowData(result)
```




